File: developer-experience/developer-experience-pwa/src/todoApi.js
import axios from "axios";
import Todo from "./model/Todo";

const apiBaseURL = "http://10.0.0.160:5000/api/v1";

export const getTodos = async () => {
  try {
    const { data } = await axios.get(`${apiBaseURL}/todos`);
    const sortedTodoItems = data.sort((el1, el2) =>
      el2.updatedAt.localeCompare(el1.updatedAt)
    );
    const todos = sortedTodoItems.map((todo) => new Todo(todo));
    return todos;
  } catch (error) {
    alert(error);
  }
};

export const getTodo = async (id) => {
  try {
    const { data } = await axios.get(`${apiBaseURL}/todos/${id}`);
    const todo = new Todo(data);
    return todo;
  } catch (error) {
    alert(error);
  }
};

export const updateTodo = async (todo, id) => {
  try {
    await axios.patch(`${apiBaseURL}/todos/${id}`, todo);
    return;
  } catch (error) {
    alert(error);
  }
};

export const addTodo = async (todo) => {
  try {
    await axios.post(`${apiBaseURL}/todos`, todo);
    return;
  } catch (error) {
    alert(error);
  }
};

export const deleteTodo = async (id) => {
  try {
    await axios.delete(`${apiBaseURL}/todos/${id}`);
    return;
  } catch (error) {
    alert(error);
  }
};

File: developer-experience/developer_experience_flutter/lib/services/todo_api.dart
import 'dart:convert';
import 'package:developer_experience_flutter/todo.dart';
import 'package:http/http.dart' as http;

class TodoApi {
  static String apiBaseUrl = 'http://10.0.0.160:5000/api/v1';

  Future<List<Todo>> getTodos() async {
    final response = await http.get(Uri.parse(apiBaseUrl + '/todos'));
    if (response.statusCode == 200) {
      Iterable l = jsonDecode(response.body);
      List<Todo> todos =
          List<Todo>.from(l.map((model) => Todo.fromJson(model)));
      return todos;
    } else {
      throw Exception('Failed to load todos');
    }
  }

  Future<Todo> getTodo(String id) async {
    final response =
        await http.get(Uri.parse(apiBaseUrl + '/todos/$id'));
    if (response.statusCode == 200) {
      Todo todo = Todo.fromJson(jsonDecode(response.body));
      return todo;
    } else {
      throw Exception('Failed to get todo');
    }
  }

  Future<void> addTodo(Todo newTodo) async {
    final response = await http.post(Uri.parse(apiBaseUrl + '/todos'),
        headers: {"Content-Type": "application/json"},
        body: jsonEncode(newTodo));
    if (response.statusCode == 201) {
      return;
    } else {
      throw Exception('Failed to post todo');
    }
  }

  Future<void> updateTodo(Todo newTodo, String id) async {
    final response = await http.patch(
        Uri.parse(apiBaseUrl + '/todos/$id'),
        headers: {"Content-Type": "application/json"},
        body: jsonEncode(newTodo));
    if (response.statusCode == 200) {
      return;
    } else {
      throw Exception('Failed to update todo');
    }
  }

  Future<void> deleteTodo(String id) async {
    final response =
        await http.delete(Uri.parse(apiBaseUrl + '/todos/$id'));
    if (response.statusCode == 200) {
      return;
    } else {
      throw Exception('Failed to delete todo');
    }
  }
}
